### DISCUSSION TITLE
Statistics and/or graphs for sensor errors  and missed reading

### DISCUSSION THREAD

**Discussion by ferringb** [NightscoutFoundation/xDrip#3678] [Ideas]
Note: I'm using version 2124041914 ; if some of these capabilities are in later versions, please let me know.  Additionally, I have xdrip+ talking directly to the G7- no intermediate dexcom app.

When you've got a semi-bad G7 it's not uncommon to have sensors errors thrown and have significant gaps in the data.  The feature request is straightforward- I think some form of configurable statistics ranges would be useful.

See ![image](https://github.com/user-attachments/assets/c526a995-1bdf-4440-a060-b5da10b77e9f) for an example of why I'm asking.   That sensor reporting is ~25% sensor errors, repeated random +50mg/dl drops, error, then bounce back to the correct value, etc.  My typical control is a stable 80-120mg/dl- but this session is obviously borked, and I'm waiting for it to throw a duration of error so I can make dexcom replace the !@#*ing thing.

Xdrip+ has the data, but AFAICT I'd have to pull logs out of it and do my own external tooling to extract this error rate.  My suggestions are thus:

* some way to access the sensor error rates- preferably as a percentile within my existing main screen stats (SD, A1c, avg, etc).  The ability to pull a historical (configurable) time range would also be useful- I'd suggest w/in the historical graph.  Knowing that yesterday was 10% errors, and today is 20%- you get the idea why I want this.
* some way to access the missed reading rates.  Xdrip+ w/ G7 seems able to backfill this when it re-establishes connectivity; I'd like better visibility into the missed readings events so I can try correlating location related issues (the actual impact of the pump on the wrong side of the body), readings progressively crapping out when I'm in a high BT environment, etc.  Beyond that, missed readings seem to be the precursor (or correlated) with sensor errors starting to occur.

I don't have time to even try hacking out this feature, and to be frank, I'm extremely reliant on my xdrip+ setup so I'm wary of experimenting with myself without some serious guard rails.  For anyone versed in the codebase, sketching out a rough 'how' for what needs to be modified, and 'how' for testing this rather than doing it 'live', that would be useful.

**Comment by jonprogers** [comment]
I don't think this covers your full requirement although I display the 'Packet capture percentage' and 'Realtime packet capture percentage' from the 'Settings>Less Common Settings>Extra Status Line' settings to give me something to watch to indicate how well my sensor session is holding up. I'm still on G6 and find that pretty reliable and normally see 100% for both those stats so if anything drops off I know to be wary of what the graph shows...

**Reply by ferringb** [reply]
Thanks, I just turned that on.  Ironically my packet capture is ~92%, realtime ~99%, so... yeah, it's just flat out sensor errors.

I don't suppose you know of a way to break the status line into multiples?  I generally have the avg/A1c/TIR; I can remove those, but I've got enough status lines active it's exceeding the screen width ;)

**Reply by jonprogers** [reply]
That's strange, mine wraps onto multiple lines in portrait and has enough space on one line in landscape. I can't see a setting to control that although I've had those display options set for so long and on multiple phones that there may something I've forgotten, sorry...

**Reply by ferringb** [reply]
@jonprogers what channel (and/or version) are your running?  Stable on my end, so I know I'm lacking the latest/greatest.  

As to looking for settings- yep, I've dug through all of that and nothing obvious has leaped out.  I've actually had this problem across my last phone and the current (pixel8 pro).

The relevant status line settings are "sliding window", "show on widget", and "calibration data".  The rest are just sources I've checked off: Average, A1c dCCT, TIR, standard deviation, packet capture percent + realtime, accuracy evaluation (which so far has no data).

**Reply by jonprogers** [reply]
I'm running the 19/9/24 nightly on a pixel 7 (xDrip master) and 27/3/24 nightly on a very old nexus 5 (xDrip follower) and both these wrap the status line. I also used to have an LG G5 and this wrapped too. I think I have been running this setup since I was on G4 so for well over 6 years and the wrapping has always happened.
I have all the things selected you list except that I have the IFCC version for A1c, plus I also have total carbs and insulin selected. I don't have any calibration shown as for G6/7 this is handled by the transmitter and not xDrip.
Copying your settings the status line still wraps, so I'm not sure what I'm doing differently, sorry.

**Reply by ferringb** [reply]
@jonprogers no worries, and thanks.  Confirming that it works for someone else points the direction for where I should go digging.

As to the G4- If I didn't start w/ G4, it was definitely early G5.  I recall having to rebuild the config *once* during that time frame- it wasn't import/export issues, there just seemed to be something fucky in it that went away when I forced a reset and rebuilt.  That's incredibly minor; no other vendor app allows me the sort of data fusion I get with xdrip+, let alone automatic uploads into tidepool.  I just wish there was a undirectional sync from tandem into xdrip+; the only thing I know of is https://github.com/jwoglom/tconnectsync , and that's tconnect -> nightscout.

That said, someone hacked out the code for this, so translating it into xdrip+ as a data source was something I kind of want to try.

**Reply by ferringb** [reply]
Oh that's just fucking weird.  Knowing it works, meant screwing around with it.  My rough repro is this:

1) keeping the phone in portrait position, add new items to the status line.  The wrapping does not occur.  This is the historical behavior I've seen, which I just assumed was a feature not yet implemented.  Note for historical- I *never* use landscape.  I detest autorotation, so it's entirely possible I've not been in xdrip w/ landscape mode basically *ever*.
2) rotate the phone into landscape.  New data items show in full.
3) rotate back into portrait: line wrapping is now occurring.

The kicker here is that the above is the exact sequence I did, and now I can't trigger it again via resetting back to previous data items that don't require a line break.  IE, trying to trigger the behavior via "no need" into "need" in portrait mode.

I can make guesses about stale configs (which I doubt- xdrip+ devs are good that way), or delayed updates (unlikely since the status line is immediately updated for new items), but I have NFC for the underlying behavior.

If the config angle is a potential- I started w/ xdrip+ late G4 or early G5, and I'm pretty much positive I had a config rebuild somewhere in late G5 era.  The last settings import/export was nov/2023 using stable from that era.  Unless others have reported something similar, I'm fine chalking this one up to "cosmic rays". ;)

**Comment by Navid200** [comment]
I've never seen a G7, or G6, that caused missed readings on and off.  I have seen a device die and show this behavior for a few last read cycles before permanently dying.  But, nothing constantly having missed readings and then recovering on and off.
  
I have seen plenty of devices set up incorrectly causing this behavior.  

What is the phone brand and model?
Do you mind showing a screenshot of the Dex Status page?

**Reply by Navid200** [reply]
If you are using a Pixel and have missed readings, I suspect that you have not disabled battery optimization properly.
If you want, we can look at that.

**Reply by ferringb** [reply]
I appreciate the offer, but this isn't PEBKAC (cellphone rather than computer, in this case).  These problems are *just for this session*, and the pump has the *same time correlated errors* in it's log, so that triangulation is just the sensor.

I'd prefer to do the sort of correlation/poking through stuff in app; I specifically want to be able to look at this when I'm out somewhere and have to decide if I can trust the sensor, or if it's better to replace it now before it deeply screws me while hiking (for example).

For missed packets- I want this for both for data geek reasons, and so I can quantify the trade off if/when I turn on the Dex "Avoid Scanning" option.  Better battery life is desired, but I'd rather quantify than guess if the trade off is worth it.

By the same angle, BT interference (noise) at high pop density events- a concert for example- is a thing.  Data  geek wise, I kind of want to see exactly how much of an impact that is.

The reason for reporting or graphing of sensor errors is obvious enough.  I've not delved into the voltage A/B relationship, but if that's useful for bug reports, same thing.

My rough thought here is to reuse the statistics implementation but just allow different data sources to be integrated; sensor error rates, missed packets, etc.  I've not dove into the code yet- my skill in java is beyond shit at this point- but if there are pointers of where to start tracing, I'd appreciate it.

**Reply by Navid200** [reply]
It will be perfectly fine to look into this and we can also make improvements to xDrip if needed.
<br/>

Google has done a poor job in the design of this page:
![Screenshot_20240922-090523](https://github.com/user-attachments/assets/9d3534ab-1bb8-4fd8-8eab-6e32621d234a)
  
There are many who look at that page and conclude that there is only one button on it.
But, the fact is that in addition to the checkbox, there is an additional button to the left of the almost invisible vertical line:
![Screenshot_20240922-090523](https://github.com/user-attachments/assets/3f233d0e-75f7-4382-a115-8eaaca9ee034)
  
Tapping on the button on the left, you will be taken to another page shown here:
![Screenshot_20240922-092158](https://github.com/user-attachments/assets/b2867660-dbc5-492b-a6a5-30b171e40633)
  
This page by default is set to optimized.  Many Pixel users don't even know this page exists.  They leave it at optimized.  And they experience intermittent connectivity.  
  
This should be set to unrestricted.  
That is half the battle.  Now, they need to do exactly the same thing for Bluetooth or legacy Bluetooth.
  
After that, Pixel will not disconnect.  
I bought my first Pixel (6a) 15 months ago and at first, experienced improved performance as far as connectivity with respect to Motorola I was used to.
I thought things had improved.  Eventually, I realized I had missed these settings for xDrip and Bluetooth. Since I have corrected the settings, I see no unexplained disconnects.  I mean, I only have a disconnect if I walk away from the phone.  

I have been using G7 for 11 months now.  I have never got a G7 sensor that has had intermittent connectivity.   
I have had sensor failures.  But, they have been obvious catastrophic failures: https://navid200.github.io/xDrip/docs/Dexcom/G7ApplicatorFailures.html

I have had one sensor that did not have any visible flaws and just did not make a transmission.  

I have never had a single G7 sensor having intermittent issues.  

Please keep us posted.  I would like to improve xDrip and am happy to do the work.
I would also like to improve my guide to help users avoid issues if possible.
So, please provide feedback.

Thanks

**Reply by ferringb** [reply]
Apologies, but can we stick to talking about the feature request I made?  Yes, my setup has unrestricted.  I have some notions about potential differences in your anecdotal experience vs mine- one large variable is my vendor being shitty and sending me near expired hardware constantly (IE, battery issues), but that is not the discussion I wish to have.  I'm here to ask for a feature and ask what the way forward is- code implementation, UI, and confirming that this is something acceptable to the maintainers of xdrip+ :)

I've not looked into the battery or voltage A/B values; I assume the G7 gives out an enumeration for the battery state ("good", "dieing soon", "dead"), and I view the A/B as safety comparisons.  When I've spotted sensor errors, the A/B are usually in strong disagreement.  I kind of want to see a graph of the variance in A/B just for geek reasons, but I understand if it's not worth the maintenance.

If either of those are are useful for bug reports, state so please.

I mentioned it earlier, but 1) I think the best UI placement is to extend the statistics section- make it possible to select different data sources, and 2) I'd appreciate if someone can point me exactly where to start making a mess in the code and trying to figure this out.  A tentative 3) is how to fully validate this (unit-tests are good, but what else?) before I try running this on my live hardware.

I've been deeply reliant on xdrip for the better of the >10y; while I dogfood a shitton of software I work with (or just flat out originated/wrote), I'm twitchy about screwing with my xdrip setup without strong guardrails.

What's the advice for the way forward?  Is there consensus that this makes sense to add as a feature, and if so, what's the process for consensus on the design/UI aspects?  I'm not asking about code review- I'm asking about who owns or decides the UI/UX side of it.

I can't say that I'll produce a usable PR, but considering the 10+ years I've used xdrip+, I should try and get over my java bias  and contribute.  Knowing exactly what I'm getting myself into- and having a pointer where to start- would be deeply appreciated.

**Reply by Navid200** [reply]
You can always open a PR.  If it solves a major problem that many suffer from, it will have to be merged.  Because if it is not, it will appear in another repository and users can always switch repositories.  
If it appears in other repositories but most users remain in this one, that means it wasn't really a major issue, or it wasn't affecting many.
Everything eventually stabilizes.  This repository is used by the majority because those maintaining it care about the community.

I only provide my opinion as a member of the community.  And I will not provide any more unsolicited input in this thread.
And sorry about the disagreement.  Nothing personal was intended.

**Reply by ferringb** [reply]
> You can always open a PR. If it solves a major problem that many suffer from, it will have to be merged. Because if it is not, it will appear in another repository and users can always switch repositories. If it appears in other repositories but most users remain in this one, that means it wasn't really a major issue, or it wasn't affecting many. Everything eventually stabilizes. This repository is used by the majority because those maintaining it care about the community.

I stick to this repository mainly for the stabilizing aspect you mentioned; it's effectively the torvalds branch in how I think of it.

I'm a complete newb when it comes to this ecosystem, and I explicitly do not want to maintain my own fork, nor cut a PR that no one will pay attention to.  What's your advice on that front?

For context, this is why I was focusing on asking about consensus; if this has no chance in hell- folks think it's a stupid idea- I'd rather find out now before putting the time in. ;)

>And I will not provide any more unsolicited input in this thread. And sorry about the disagreement. Nothing personal was intended.

Online commenting sucks for not sounding like a raging asshole, which I probably did.  No offense was taken, and having maintained a lot of communities, I grok your intent- and I've absolutely done the same thing.  I just was trying to skip you forward to giving me advice from the developmental side of this so I could contribute.

Any advice you've got there- solicited/unsolicited, whichever- I'd appreciate.

Also, I'm aparently incompetent at writing proper github comments.  Apologies for the quoting mistake. ;)

**Reply by Navid200** [reply]
I hope we can first convince both of us that you are getting the best you can get from xDrip as is now.

I believe that you already believe that.  If you give me a chance to convince myself in that respect as well, we can then talk about what you are thinking of with regards to changing xDrip so that it would provide more details about capture rate and the best way in doing that.  And you are not the only one who wants improvements there.

I don't want to brag.  But, my capture rate is 100%.  If you had a garbage Samsung, I wouldn't have much to say.  But, you are using a Pixel and your capture rate is less than 99%?  That makes no sense.  Something is definitely wrong there and I would like to find it.

Now, you say why can we not talk just about adding a feature and have to talk about troubleshooting?
I cannot advise you if it has no chance of being accepted.  The first question asked after you open a PR is why do we need this?
I don't know the answer to that question right now.  So, I cannot advise you unless I am sure something is really wrong with xDrip as is.

**Reply by Navid200** [reply]
This is an existing PR that adds capture rate to statistics:
https://github.com/NightscoutFoundation/xDrip/pull/2698

I would like to clean up the statistics before we add this.  I will.
But, it takes time and there is so much going on.  

Every time Google changes something, lots of things fall apart and we have to fix them.  
xDrip sync is a feature that allows parents follow their children.  It sopped working because a change Google made 2 months ago and our lead developer spent a lot of time coming up with a solution.  That took quite a while.
We are now waiting to make sure it has no issues and we are monitoring how it gradually gets used by everyone as the old scheme that was broken by Google is gradually fading out.  In the meantime, PRs are all frozen!

None of this is visible to everyone.  Every new feature makes this even more complicated.  

There is a request for offering audio focus for spoken readings.  We need to add that.
There is a request for offering customized vibration patterns for alarms.  We need to offer that.
New devices come out and we need to support them.
There are requests for new insulin types. There are requests about WearOS.  There are requests for allowing calibrations in so many different weird setups.
There is no end to requests. 

All of these take time and we prioritize them and address them all.
But, we have to ask questions and go over every proposal and discuss and sometimes the way forward may not make everyone happy.   But, I don't know any other way.
